### PR TITLE
IRSymTab: Record _GLOBAL_OFFSET_TABLE_ for ELF x86

### LINKED ISSUE #61101: lld error "undefined symbol: _GLOBAL_OFFSET_TABLE_ " with LTO
The following test file fails with LTO when compiled for x86. Other arches or bfd don't seem to exhibit this.

```console
$ cat clang-bug.c 
void a(void) {}
void _start(void) { a(); }

$ clang -target i386-pc-linux-gnu -fuse-ld=lld -nostdlib -o /dev/null clang-bug.c

$ clang -target i386-pc-linux-gnu -fuse-ld=lld -nostdlib -o /dev/null clang-bug.c -flto
ld.lld: error: undefined symbol: _GLOBAL_OFFSET_TABLE_
>>> referenced by ld-temp.o
>>>               lto.tmp:(_start)
clang-15: error: linker command failed with exit code 1 (use -v to see invocation)
```

### PR DESCRIPTION
**Pull request by MaskRay**
In ELF, relocatable files generated for x86-32 and some code models of
x86-64 (medium, large) may reference the special symbol
`_GLOBAL_OFFSET_TABLE_` that is not used in the IR. In an LTO link, if
there is no regular relocatable file referencing the special symbol, the
linker may not define the symbol and lead to a spurious "undefined
symbol" error.

Fix #61101: record that `_GLOBAL_OFFSET_TABLE_` is used in the IR symbol
table.

Note: The `PreservedSymbols` mechanism
(https://reviews.llvm.org/D112595) that just sets `FB_used` is not
applicable.
The `getRuntimeLibcallSymbols` for extracting lazy runtime library
symbols is for symbols that are "always" potentially used, but linkers
don't have the code model information to make a precise decision.

### CODE REVIEW

File: lld/test/ELF/lto/i386-global-offset-table.ll [hunk 1]
```diff
@@ -0,0 +1,22 @@
+; REQUIRES: x86
```
Comment (teresajohnson) [inline lld/test/ELF/lto/i386-global-offset-table.ll:1]: What is this test testing?

File: lld/test/ELF/lto/x86-64-global-offset-table.ll [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+; RUN: cat a.ll large.ll | llvm-as - -o large.bc
+; RUN: ld.lld -pie large.bc b.o -o large
+; RUN: llvm-objdump -d large | FileCheck %s
+
+; RUN: cat a.ll medium.ll ref.ll | llvm-as - -o ref.bc
+; RUN: ld.lld -pie --no-relax -u ref ref.bc b.o -o ref
+; RUN: llvm-objdump -d ref | FileCheck %s
+
+; CHECK: movabsq
```
Comment (teresajohnson) [inline lld/test/ELF/lto/x86-64-global-offset-table.ll:17]: What is this checking for related to the _GLOBAL_OFFSET_TABLE_? Can you add a comment?

File: llvm/test/LTO/X86/codemodel-2.ll [hunk 3]
```diff
@@ -1,5 +1,5 @@
 ; RUN: llvm-as %s -o %t.o
-; RUN: llvm-lto2 run -r %t.o,_start,px %t.o -o %t.s
+; RUN: llvm-lto2 run -r %t.o,_start,px -r %t.o,_GLOBAL_OFFSET_TABLE_, %t.o -o %t.s
```
Comment (teresajohnson) [inline llvm/test/LTO/X86/codemodel-2.ll:2]: Why is it needed here and in other tests now?

**Review by teresajohnson — COMMENTED** [review]
A few questions about tests below. Also, ModuleSymbolTable::CollectAsmSymbols is called when building the ModuleSummaryIndex for ThinLTO - I don't see any ThinLTO tests modified or added here - do we need one to make sure nothing surprising happens in that mode?

**Comment by aeubanks** [comment]
I was trying to understand why we couldn't resolve `_GLOBAL_OFFSET_TABLE_` after LTO in lld. Ideally we wouldn't special case linker-synthesized symbols in LLVM?

File: lld/test/ELF/lto/i386-global-offset-table.ll (hunk 1 above)
Comment (MaskRay) [reply lld/test/ELF/lto/i386-global-offset-table.ll:1]: Thanks for the review. Added a file-level comment.

File: llvm/test/LTO/X86/codemodel-2.ll (hunk 3 above)
Comment (MaskRay) [reply llvm/test/LTO/X86/codemodel-2.ll:2]: For x86-64 medium/large code model, the IR symbol table now references `_GLOBAL_OFFSET_TABLE_`, which requires llvm-lto2 to provide it to avoid a ` missing symbol resolution for ` error

**Comment by MaskRay** [comment]
> I was trying to understand why we couldn't resolve `_GLOBAL_OFFSET_TABLE_` after LTO in lld. Ideally we wouldn't special case linker-synthesized symbols in LLVM?

lld only defines `_GLOBAL_OFFSET_TABLE_`  when it is referenced. The symbol table should be finalized before LTO.

Codegen passes may reference a runtime library function or linker-defined symbols `_GLOBAL_OFFSET_TABLE_` that is not in the referencer's symbol table. In such a case, if we don't "workaround" it, we will get an "undefined symbol" error.
This behavior is not lld specific and applies to other linkers). I feel that LTO is the right place to implement the special logic. If we have other symbols that are not ELF-specific, a linker-oriented solution would require to patch every linker.

**Comment by aeubanks** [comment]
Linker-defined symbols seem different from runtime library symbols.

Runtime library symbols have to be provided by some input object file. We can either do something similar to this PR and mark all LTO bitcode files as requiring any potential runtime library symbol (there are a lot, so it might take up noticeable space in the bitcode files), or have the linker special case them (which is what I believe lld does).

But for linker-defined symbols, it seems like it's a fixable phase ordering issue by defining them after parsing object files created by LTO. Then we don't have to start special casing any of these types of symbols. Perhaps I'm missing something though.

**Comment by MaskRay** [comment]
We want the symbol table to be complete before `compileBitcodeFiles<ELFT>(skipLinkedOutput);`. In the false positive cases, `_GLOBAL_OFFSET_TABLE_` was not in the symbol table, so lld did not know the symbol would later be used by generated relocatable files. This cannot be fixed by reordering some passes in the linker.

This could be solved by making the linker try probing `_GLOBAL_OFFSET_TABLE_` after LTO compilation, but it doesn't feel right to me...

For runtime library symbols, ideally we the solution should be in LLVMLTO instead of linkers. However, given that we have to be very conservative, that'd cause a lot of overhead in bitcode files, so we accepted to workaround in linkers.

**Comment by aeubanks** [comment]
(fwiw I think this change is an acceptable alternative, especially if the number of linker-defined symbols across all architectures that we need to special case is small)

**Comment by pcc** [comment]
The intent was always that LTO would tell the linker which symbols may be used by the combined object file, and that it wouldn't be handled by the individual bitcode files. See `LTO::getRuntimeLibcallSymbols()` which is where the list currently comes from (it currently refers to "library functions" but I don't see why it should be limited to that). So I am not sure that `_GLOBAL_OFFSET_TABLE_` should be a special case here.

**Comment by pcc** [comment]
> the difference is that for runtime library calls, adding all potentially created runtime calls to the bitcode symbol table is expensive.

More importantly, there would be no purpose in embedding the list in every object file. The bitcode symbol table is like a cache for the information in the rest of the bitcode file, and the behavior of the LLVM backend linked into lld is what ultimately matters for LTO because that's what generates library calls. Since the list is the same for every bitcode file, and it would go out of date if producer version != consumer version, the "cache" in the individual bitcode files would serve no purpose because `LTO::getRuntimeLibcallSymbols()` would always provide you with the correct symbol list.

**Comment by MaskRay** [comment]
> The intent was always that LTO would tell the linker which symbols may be used by the combined object file, and that it wouldn't be handled by the individual bitcode files. See `LTO::getRuntimeLibcallSymbols()` which is where the list currently comes from (it currently refers to "library functions" but I don't see why it should be limited to that). So I am not sure that `_GLOBAL_OFFSET_TABLE_` should be a special case here.

The commit message states:

> The PreservedSymbols mechanism
([reviews.llvm.org/D112595](https://reviews.llvm.org/D112595)) that just sets FB_used is not
applicable.

That mechanism and `LTO::getRuntimeLibcallSymbols()`  handle lazy symbols that appear in the symbol table before LTO compilation. `_GLOBAL_OFFSET_TABLE_` is a special case that it may not be in the symbol table when LTO compilation occurs.

Technically linkers can add the special case to `handleLibcall` by ensuring `_GLOBAL_OFFSET_TABLE_` is always present (thigh might only be needed for x86). Personally I feel that LLVMLTO, which is equipped with more information, is the best place to implement this to benefit all linkers (GNU ld might always define `_GLOBAL_OFFSET_TABLE_` when dynamic sections are needed).

**Comment by pcc** [comment]
> That mechanism and `LTO::getRuntimeLibcallSymbols()` handle lazy symbols that appear in the symbol table before LTO compilation. `_GLOBAL_OFFSET_TABLE_` is a special case that it may not be in the symbol table when LTO compilation occurs.

Isn't that just an implementation detail of lld? Basically, the behavior of `handleLibcall` should be as if we loaded an object file with an undefined symbol with that name. Currently we only implement the archive loading aspect of that behavior and only if the target is a bitcode file as a workaround for the libgcc issue, but maybe what we should really be doing is actually defining the symbols via `addSymbol` (unless there's a non-bitcode archive member to keep working around the libgcc issue)?

> Technically linkers can add the special case to `handleLibcall` by ensuring `_GLOBAL_OFFSET_TABLE_` is always present (thigh might only be needed for x86). Personally I feel that LLVMLTO, which is equipped with more information, is the best place to implement this to benefit all linkers (GNU ld might always define `_GLOBAL_OFFSET_TABLE_` when dynamic sections are needed).

Right, LTO (on behalf of the LLVM backend) should be the source of truth of this information, and it can be accessed via `LTO::getRuntimeLibcallSymbols()`. It doesn't follow that this information needs to be in every bitcode symbol table.

**Comment by MaskRay** [comment]
> > That mechanism and `LTO::getRuntimeLibcallSymbols()` handle lazy symbols that appear in the symbol table before LTO compilation. `_GLOBAL_OFFSET_TABLE_` is a special case that it may not be in the symbol table when LTO compilation occurs.
> 
> Isn't that just an implementation detail of lld? Basically, the behavior of `handleLibcall` should be as if we loaded an object file with an undefined symbol with that name. Currently we only implement the archive loading aspect of that behavior and only if the target is a bitcode file as a workaround for the libgcc issue, but maybe what we should really be doing is actually creating the undefined symbols via `addSymbol(Undefined{...})` (unless there's a non-bitcode archive member to keep working around the libgcc issue)?

I think it's related to an implementation detail of lld and other linkers, but more related to the LTO symbol resolution scheme. The IR symbol table should strive to provide a full picture. In this case, it knows the architecture and the code model and can make a best guess of whether `_GLOBAL_OFFSET_TABLE_` will be needed.

> > Technically linkers can add the special case to `handleLibcall` by ensuring `_GLOBAL_OFFSET_TABLE_` is always present (thigh might only be needed for x86). Personally I feel that LLVMLTO, which is equipped with more information, is the best place to implement this to benefit all linkers (GNU ld might always define `_GLOBAL_OFFSET_TABLE_` when dynamic sections are needed).
> 
> Right, LTO (on behalf of the LLVM backend) should be the source of truth of this information, and it can be accessed via `LTO::getRuntimeLibcallSymbols()`. It doesn't follow that this information needs to be in every bitcode symbol table.

I hope that the linker only defines `_GLOBAL_OFFSET_TABLE_` when it is really used.
At the `LTO::getRuntimeLibcallSymbols()` call site there is no information about code models to make a good decision.

The IRSymTab code will likely be extended to be smarter on function-local inline asm symbols (beside current model-level inline asm symbols).
Adding `_GLOBAL_OFFSET_TABLE_` there feel natural to me.

**Comment by pcc** [comment]
Okay, that makes sense. `LTO::getRuntimeLibcallSymbols()` should be used for symbols that are "always" potentially used. We could extend that notion to "always potentially used on a given architecture" by passing in an architecture (which the linker knows), but the linker doesn't know the code model. So adding it to the bitcode symbol table seems fine to me.

File: lld/test/ELF/lto/x86-64-global-offset-table.ll [hunk 4]
```diff
@@ -0,0 +1,71 @@
+; REQUIRES: x86
+;; LTO-generated felocatable files may reference _GLOBAL_OFFSET_TABLE_ while
```
Comment (rnk) [inline lld/test/ELF/lto/x86-64-global-offset-table.ll:2]: s/felocat/relocat/

**Review by aeubanks — APPROVED** [review]

File: lld/test/ELF/lto/x86-64-global-offset-table.ll (hunk 4 above)
Comment (MaskRay) [reply lld/test/ELF/lto/x86-64-global-offset-table.ll]: Thanks. Fixed

File: lld/test/ELF/lto/x86-64-global-offset-table.ll (hunk 2 above)
Comment (MaskRay) [reply lld/test/ELF/lto/x86-64-global-offset-table.ll:31]: Added `The IR symbol table references _GLOBAL_OFFSET_TABLE_, which causes lld to define the symbol.`

**Review by teresajohnson — APPROVED** [review]
lgtm with one comment fix below.
